File: backend/src/main/java/com/resulthub/api/complaint/entity/Complaint.java
package com.resulthub.api.complaint.entity;

import com.resulthub.api.user.User;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "complaints")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Complaint {

    @Id
    @GeneratedValue
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creator_id")
    private User creator;

    @Column(nullable = false, length = 100)
    private String category;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(name = "media_urls", columnDefinition = "text[]")
    private String[] mediaUrls;

    @Column(precision = 10, scale = 7)
    private BigDecimal latitude;

    @Column(precision = 10, scale = 7)
    private BigDecimal longitude;

    @Column(name = "location_name", length = 255)
    private String locationName;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    @Builder.Default
    private ComplaintStatus status = ComplaintStatus.OPEN;

    @Column(name = "is_anonymous")
    @Builder.Default
    private Boolean isAnonymous = false;

    @Column(name = "flag_count")
    @Builder.Default
    private Integer flagCount = 0;

    @Builder.Default
    private Integer upvotes = 0;

    @Builder.Default
    private Integer downvotes = 0;

    @Column(name = "net_score", insertable = false, updatable = false)
    private Integer netScore;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum ComplaintStatus {
        OPEN,
        UNDER_REVIEW,
        RESOLVED
    }
}

File: backend-mern/src/models/Complaint.ts
import mongoose, { Schema, Document } from 'mongoose';

export enum ComplaintStatus {
  OPEN = 'OPEN',
  IN_PROGRESS = 'IN_PROGRESS',
  RESOLVED = 'RESOLVED',
  CLOSED = 'CLOSED'
}

export enum ComplaintPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

export interface IComplaint extends Document {
  creator?: mongoose.Types.ObjectId;
  workspace?: mongoose.Types.ObjectId;
  category: string;
  title: string;
  description: string;
  mediaUrls: string[];
  latitude?: number;
  longitude?: number;
  locationName?: string;
  status: ComplaintStatus;
  priority: ComplaintPriority;
  assignee?: mongoose.Types.ObjectId;
  anonymous: boolean;
  flagCount: number;
  upvotes: number;
  downvotes: number;
  netScore: number;
  createdAt: Date;
  updatedAt: Date;
}

const ComplaintSchema: Schema = new Schema({
  creator: { type: Schema.Types.ObjectId, ref: 'User' },
  workspace: { type: Schema.Types.ObjectId, ref: 'Workspace' },
  category: { type: String, required: true, maxlength: 100 },
  title: { type: String, required: true, maxlength: 255 },
  description: { type: String, required: true },
  mediaUrls: [{ type: String }],
  latitude: { type: Number },
  longitude: { type: Number },
  locationName: { type: String },
  status: { type: String, enum: Object.values(ComplaintStatus), default: ComplaintStatus.OPEN },
  priority: { type: String, enum: Object.values(ComplaintPriority), default: ComplaintPriority.MEDIUM },
  assignee: { type: Schema.Types.ObjectId, ref: 'User' },
  anonymous: { type: Boolean, default: false },
  flagCount: { type: Number, default: 0 },
  upvotes: { type: Number, default: 0 },
  downvotes: { type: Number, default: 0 },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for netScore
ComplaintSchema.virtual('netScore').get(function() {
  return this.upvotes - this.downvotes;
});

export default mongoose.model<IComplaint>('Complaint', ComplaintSchema);
